src/pages/index.js

```diff
@@ -63,27 +102,110 @@ function handleImageClick(cardLink, cardCaption) {
   imagePopup.open(cardLink, cardCaption);
 }
 
+function handleAvatarClick() {
+  editAvatarPopup.open();
+}
+
+function handleDeletePopup(cardId, evt) {
+  deleteVerifyPopup.open();
+  cardIdForDeletion = cardId;
+  cardToDelete = evt.target.closest(".card");
+}
+
+function likeToServer(cardId, cardElement) {
+  if (!cardElement.querySelector(".card__liked_active")) {
+    api.removeLikeFromCard(cardId).catch((err) => console.error(err));
+  } else {
+    api.addLikeToCard(cardId).catch((err) => console.error(err));
+  }
+}
+
 function createCard(imageInfo) {
-  const newCard = new Card(imageInfo, "#card-template", handleImageClick);
+  const newCard = new Card(
+    imageInfo,
+    "#card-template",
+    handleImageClick,
+    handleDeletePopup,
+    likeToServer
+  );
   const cardElement = newCard.generateCard();
   return cardElement;
 }
 
+function renderLoading(isLoading, button) {
+  if (isLoading) {
+    buttonText = button.textContent;
+    button.textContent = "Saving...";
+  } else {
+    button.textContent = buttonText;
+  }
+}
+
 function handleAddImage(imageInfo) {
-  const addedCardElement = createCard(imageInfo);
-  cardSection.addItem(addedCardElement, "prepend");
-  addFormPopup.close();
-  addFormPopup.clearForm();
-  formValidators["add"].disableSubmitButton();
+  renderLoading(true, addPopup.querySelector(config.submitButtonSelector));
+  api
+    .addCard(imageInfo)
+    .then((res) => {
+      let addedCardElement = createCard(res);
+      cardSection.addItem(addedCardElement, "prepend");
+    })
+    .catch((err) => console.error(err))
+    .finally(() => {
+      renderLoading(false, addPopup.querySelector(config.submitButtonSelector));
+      addFormPopup.close();
+      addFormPopup.clearForm();
+      formValidators["add"].disableSubmitButton();
+    });
 }
 
 function handleEditProfile(newProfileInfo) {
-  userInfo.setUserInfo(newProfileInfo);
-  editFormPopup.close();
+  renderLoading(true, editPopup.querySelector(config.submitButtonSelector));
+  userInfo.changeUserInfo(newProfileInfo);
```

Comment:
You need to update `DOM` only in blocks `then` after a successful response from the server so that when there is an error it wouldn’t change